core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -492,6 +453,71 @@ private[kafka] class Processor(val id: Int,
     }
   }
 
+  /* `protected` for test usage */
+  protected[network] def sendResponse(response: RequestChannel.Response) {
+    trace(s"Socket server received response to send, registering for write and sending data: $response")
+    val channel = selector.channel(response.responseSend.destination)
+    // `channel` can be null if the selector closed the connection because it was idle for too long
```

Comment:
I think you're talking about the case where `key.isValid` returns false during `poll`. In my tests, closing the client connection did not trigger that. Looking at the docs, it says:

```html
 <p> A key is valid upon creation and remains so until it is cancelled, its channel is closed, or its selector is closed.  </p>
```

So, it's not clear to me that this actually ever returns `false` if the client connection is closed.

From the reading I have done, it seems that an actual operation on the channel (read or write) is required to detect client closes. But it was hard to find authoritative information on this.

For the the cases where the channel is no longer valid, but it's still in the `Selector`, calling `Selector.send` will not throw an exception, it will record the failed send as a disconnection and other code will take care of that.

Let me know if you are happy with the code as it is or if you'd prefer if I change the comment to say instead:

```java
// `channel` can be null if the selector closed the connection (eg because it was idle for too long)
```